Update Query documentation

smitpatel opened the pull request:
Resolves#615
Resolves#1636
Resolves#1648
Resolves#1699
Resolves#1777

AndriySvyryd commented on .openpublishing.redirection.json:
```diff
@@ -214,6 +214,17 @@
+            "redirect_url": "/ef/core/querying/index",
+            "redirect_document_id": false
+        }
+        ,
```
Extra newline

AndriySvyryd commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+
+## Potential memory leak in client evaluation
+
+Since query translation and compilation is expensive, by default, EF Core caches compiled query plan. With client evaluation allowed in the top projection, the cached delegate can use client code. In most cases, EF Core generates parameters for client parts of the tree so that their values can be replaced when reusing the query plan. But with certain constants in expression tree, it cannot be con ...
```
```suggestion
Since query translation and compilation is expensive, by default, EF Core caches the compiled query plan. With client evaluation allowed in the top projection, the cached delegate can use client code. In most cases, EF Core generates parameters for client parts of the tree so that their values can be replaced when reusing the query plan. But certain constants in expression tree cannot be converted to parameters. If the client code contains such constants then those objects cannot be garbage collected as they would still be referenced from the cached query plan. If such object contains DbContext or other services inside it then it could cause memory usage of the app to grow over time. In order to avoid potential memory leak, EF Core throws an exception whenever it encounters constants of type which cannot be mapped using the current database provider. Common causes and their solutions are as follows:
```

AndriySvyryd commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+---
+# Complex Query Operators
+
+Language Integrated Query (LINQ) contains many complex operators which combines multiple data sources or does complex processing. Not all LINQ operators have suitable translation in server side. Sometimes, query written in one form can be translated but in different form doing different computation cannot be translated. This page describes some of the complex operators and their supported variati ...
```
```suggestion
Language Integrated Query (LINQ) contains many complex operators which combines multiple data sources or does complex processing. Not all LINQ operators have suitable translations on the server side. Sometimes, query written in one form can be translated but in different form it can't be even if the result is same. This page describes some of the complex operators and their supported variations. In future we may recognize more patterns and add their translations. Important point to remember is that database capabilities also vary, so a particular query which is translated on SqlServer may not work for a Sqlite database.
```

bricelam replied:
Casing should be *SQLite* unless you're referring to our packages or APIs.

AndriySvyryd commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+
+## GroupJoin
+
+LINQ GroupJoin operator allows you to connect two data sources in same way as Join but the different here is, it creates a group of values from inner source which matches key with element from outer source. Performing query like following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) do not support representing a collection of client side obje ...
```
```suggestion
LINQ GroupJoin operator allows you to connect two data sources in same way as Join but the difference here is that it creates a group of values from inner source which matches the key with an element from the outer source. Performing a query like the following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) do not support representing a collection of client side objects, GroupJoin cannot be translated to server in many cases. It requires to get all the data from server to perform GroupJoin without a special selector. But if the selector is limiting data being selected then fetching all data from server may cause performance issues. Hence EF Core does not translate GroupJoin.
```

AndriySvyryd commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+
+## SelectMany
+
+LINQ SelectMany operator allows you to enumerate over collection selector for each element in outer source and generate tuples of value from each data source. In a way it is a join but without any condition so every record is connected with other. Depending on how collection selector is related to outer data source, SelectMany can translate to various different queries on server side.
```
```suggestion
LINQ SelectMany operator allows you to enumerate over collection selector for each element in outer source and generate tuples of value from each data source. In a way it is a join but without any condition so every record is connected with another. Depending on how collection selector is related to outer data source, SelectMany can translate to various different queries on server side.
```

AndriySvyryd commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+
+### Collection selector does not reference outer
+
+When collection selector is not referencing anything from outer source, it is cartesian product of both data sources. In translates to `CROSS JOIN` in relational databases.
```
```suggestion
When collection selector is not referencing anything from the outer source, it is a cartesian product of both data sources. In translates to `CROSS JOIN` in relational databases.
```

AndriySvyryd commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+
+### Collection selector references outer in predicate
+
+When collection selector has a where predicate which is using element from outer source, normally this arises when using collection navigation on outer source as collection selector, then such where predicate can be used as join condition and it can translate to join on databases. If the collection is empty for outer element, then it would not generate any result for that outer element. But if `D ...
```
```suggestion
When collection selector has a where predicate which is using an element from the outer source (normally this arises when using collection navigation on outer source as collection selector) then such where predicate can be used as the join condition and it can be translated to a join on the server side. If the collection is empty for the outer element, then it would not generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with the default value of inner element. Due to this distinction, this kind of queries translate to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

bricelam commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+
+## Potential memory leak in client evaluation
+
+Since query translation and compilation is expensive, by default, EF Core caches compiled query plan. With client evaluation allowed in the top projection, the cached delegate can use client code. In most cases, EF Core generates parameters for client parts of the tree so that their values can be replaced when reusing the query plan. But with certain constants in expression tree, it cannot be con ...
```
I wonder if we should avoid the word *constant* here.

smitpatel replied:
What should we call it?

smitpatel replied:
ideas?

bricelam replied:
Not sure, but I think to most users would think of the `const` keyword.
*references to fields and properties* gets a little closer, but probably isn't right

smitpatel replied:
reference to "this" is the only one which causes issue.

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,120 @@
+
+## Join
+
+LINQ Join operator allows you to connect two data sources based on the key selector on each source and generates a tuple of values when key matches. This naturally translates to `INNER JOIN` on relational databases. While LINQ Join has outer & inner key selectors, databases take a condition as join condition and EF Core will generate join condition by comparing outer key selector to inner key sel ...
```
Would *left* and *right* be clearer than *outer* and *inner*?

bricelam replied:
*source* might also be confusing. Maybe *set* is more idiomatic for EF users

smitpatel replied:
Open to ideas. APIs call them outer & inner source.

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,120 @@
+CROSS JOIN [Posts] AS [p]
+```
+
+### Collection selector references outer in predicate
````
Another word that might be too technical: *predicate*

smitpatel replied:
where? filter?

bricelam replied:
where clause?

bricelam commented on entity-framework/core/querying/how-query-works.md:
```diff
@@ -3,7 +3,7 @@ title: How Queries Work - EF Core
 ms.date: 09/26/2018
 ms.assetid: de2e34cd-659b-4cab-b5ed-7a979c6bf120
-uid: core/querying/overview
+uid: core/querying/how-query-works
```
I don't think you can change these. Was the old one redirected?

smitpatel replied:
yes. I added redirects.

smitpatel commented on entity-framework/core/querying/async.md:
```diff
@@ -1,14 +1,14 @@
-author: rowanmiller
-ms.date: 01/24/2017
+author: smitpatel
+ms.date: 10/01/2019
```
Removed

smitpatel replied:
[Warning] Missing attribute: ms.date. A freshness date is required for this content set. Add a date in format MM/DD/YYYY.

need to add it. :X

AndriySvyryd replied:
Also I remember us discussing that it should represent the date when there was a content change, not just formatting or spelling

smitpatel commented:
Updated.

roji commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+### Collection selector references outer in where clause
+
+When collection selector has a where predicate, which is referencing the outer element, then EF Core translates it to join on database and uses the predicate as join condition. Normally it arises when using collection navigation on outer source as collection selector. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` ...
```
```suggestion
When the collection selector has a Where predicate which references the outer element, EF Core translates it to a database join and uses the predicate as the join condition. Normally this occurs when using a collection navigation on an outer source as the collection selector. If the collection is empty for the outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector, the outer element will be connected with the default value of the inner element type. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

smitpatel replied:
Acrolinx disallows uses of this.

roji replied:
That is not a valid response to a review comment. This fragment is clearly missing articles.

roji commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+ms.assetid: 2e187a2a-4072-4198-9040-aaad68e424fd
+uid: core/querying/complex-query-operators
+---
+# Complex Query Operators
```
We probably need a note that this page applies only to relational database.

smitpatel replied:
That is incorrect.

roji replied:
Care to elaborate?

smitpatel replied:
The page is not necessarily put here for relational. We can add more operators here as we go. Like Cosmos can translate bunch of SelectMany in awesome way. The main purpose in not putting relational here is to reuse this page as needed for other providers in future. Further, all the operators talk about the relational translation so it is evident that it applies to relational only.

roji replied:
Re Cosmos translations specifically, we may want to consider talking about that in the Cosmos provider docs - not sure it's a good idea to pack all the provider-specific info in one page. It will not be relevant for most readers of this page. Otherwise the page as it is seems 100% relational to me, but I don't have super strong feelings on this.

roji replied:
Yeah, there is one point on this long and elaborate page that is also common to non-relational providers.

As I said if you feel very strongly about not including a note on this, I definitely won't die on that hill.

roji commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+### Collection selector references outer in non-predicate
+
+When collection selector has reference to outer, which isn't in predicate (as the above case), it can't be converted to join. That's why we need to evaluate collection selector for each outer element. It translates to `APPLY` operation in relational databases. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is appl ...
```
This is quite SQL Server-specific.

```suggestion
When the collection selector references the outer element, but from a non-predicate operator (as was the  case above), the query can't be translated to a regular database join, since we need to evaluate collection selector for each outer element. However, the query does translate to an `APPLY` or a lateral join operation in relational databases.

As before, if the collection is empty for the outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with the default value of inner element type. Because of this distinction, on SQL Server this kind of query translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. 

Sqlite doesn't support `APPLY` operator, and therefore this kind of query cannot be translated when targeting it.
```

smitpatel replied:
Lateral join and apply operations are different. It is fine the way it is. It is Oracle-specific.

roji replied:
They're not really that different. PostgreSQL implements the APPLY behavior with a LATERAL JOIN, there's just an additional ON TRUE there.

smitpatel replied:
Again Lateral joins are apply operations are different. Postgre fits apply operation on lateral join with additional condition. I have added statement that query may not translate always from EF docs perspective, we generate Apply operation and that is accurate doc content.

roji replied:
The fact is that what this section describes (Collection selector references outer in non-predicate) is translated in PostgreSQL via lateral join, which is part of the SQL standard; PostgreSQL users reading this will have no idea about APPLY. All I'm asking is to mention lateral join alongside APPLY so that they understand, because that's what is used for translation.

smitpatel replied:
That is misleading. Relational generates Apply operation. EF Core relational layer does not generate lateral joins. Writing lateral join easily confuses user asking, oh why did not translate this query via lateral join. Sorry but we don't support lateral join.

roji replied:
I think I'll let this PR go now.

roji commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## GroupBy
+
+LINQ GroupBy operator creates result of type `IGrouping<TKey, TElement>` where `TKey` & `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over using any LINQ operator after grouping. Since no database structure can represent IGrouping, Group By operator has no translation in most cases. When an aggregate operator is applied t ...
```
```suggestion
The LINQ GroupBy operator returns results of type `IGrouping<TKey, TElement>`, where `TKey` and `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means it can be composed over using any LINQ operator after grouping. Since no database structure can represent `IGrouping`, the GroupBy operator cannot be translated in most cases. As an exception, when an aggregate operator is applied to each group (e.g. MAX, COUNT), the query can be translated to  SQL `GROUP BY`. 

Note that SQL `GROUP BY` is restrictive too, requiring you to group only by scalar values; it also requires the projection to only contain grouping key columns or applied aggregate operators. EF Core identifies this subset and translates it to server, as in the following example:
```

smitpatel replied:
Acrolinx doesn't like long statements.

smitpatel replied:
Acrolinx believe that Note that does not add any value and unncessary.

roji commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+GROUP BY [p].[AuthorId]
+```
+
+EF Core also translates when aggregate operator appears in Where filter by using `HAVING` clause in SQL and aggregate operator appears in Order By (or other ordering operators). The query before applying Group By operator can be any complex query using various different operators as long as it can be translated to server. Further, once you apply aggregate operator on grouping query to remove grou ...
````
```suggestion
EF Core can also translate queries where an aggregate operator appears in a Where filter (by generating an SQL `HAVING` clause), or in an OrderBy clause (or other ordering operators). The query part before the GroupBy operator can be any complex query as long as it can be translated to server. Further, once you apply aggregate operator on a grouping query as shown above, you can compose it like any other query.
```

smitpatel replied:
Avoid usage of can when we do it.

smitpatel replied:
Acrolinx also dislikes parenthesis.

roji commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+- Min
+- Sum
+
+## Left Join
```
This repeats the above GroupJoin section no?

smitpatel replied:
Nope.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientMethod)]
 
-## Client evaluation performance issues
+## Unsupported client evaluation
```
I don't agree with this reasoning. A client code worked in projection failed in where. Users would have question why that happens.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+
+User may need to force into client evaluation explicitly in certain cases like following
+
+- Data is small so that doing evaluation on client doesn't cause huge performance penalty.
```
You can disagree with it but our users don't. Our users are indeed using `AsEnumerable` to cause client eval and it is part of our exception message too.

roji replied:
I'm talking about your logic. The fact that something doesn't cause a huge perf penalty is not, in itself, a justification to use it. You use it because you have no choice or because there's some other benefit.

smitpatel replied:
Again that is not true. It is simple cost vs benefit analysis. If I have table of 40 rows and a client filter function, I can also convert the client function to a database function and use UDF to map it. I can get best performance at the expense of spending ours to generate UDF when I can just do client eval because it is just cheaper and does not cause huge penalty.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+
+## Potential memory leak in client evaluation
+
+Since query translation and compilation are expensive, EF Core caches the compiled query plan. The cached delegate may use client code while doing client evaluation of top-level projection. EF Core generates parameters for client parts of the tree, to reuse the query plan by replacing parameter values. But certain constants in expression tree can be converted to parameters. If the cached delegate ...
```
I don't think users are ever going to understand it anyway. :trollface: 

Important point is 3 solutions which describes what to do when we throw exception.

roji replied:
This needs simple code samples instead of a wall of text.

smitpatel commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+INNER JOIN [Person] AS [p] ON [p0].[PersonPhotoId] = [p].[PhotoId]
+```
+
+## GroupJoin
````
No variant of GroupJoin is translatable. Period.
The 2 queries are to demonstrate different amount of data needed to be fetch from server.
There is very little confusion about distinction between join/groupjoin syntax in customers. Most customers so far have known difference. The other query is not `from.... where...` it is actual group join and then composition afterwards. So I don't see any clarification required.

smitpatel commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+When collection selector isn't referencing anything from the outer source, the result is a cartesian product of both data sources. It translates to `CROSS JOIN` in relational databases.
+
+[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToCrossJoin)]
```
Decided not to do. LINQ has both syntax for various thing and LINQ docs can describe that. Our docs would be really long with very low content relating to us.

smitpatel commented on entity-framework/core/querying/async.md:
```diff
@@ -1,21 +1,21 @@
 # Asynchronous Queries
 
-Asynchronous queries avoid blocking a thread while the query is executed in the database. This can be useful to avoid freezing the UI of a thick-client application. Asynchronous operations can also increase throughput in a web application, where the thread can be freed up to service other requests while the database operation completes. For more information, see [Asynchronous Programming in C#](h ...
+Asynchronous queries avoid blocking a thread while the query is executed in the database. It avoids freezing the UI of a thick-client application. Async operations can also increase throughput in a web application, where async query would free up the thread to service other requests while the database operation completes. For more information, see [Asynchronous Programming in C#](/dotnet/csharp/a ...
```
Acrolinx score went down by 2.

smitpatel replied:
I changed it something more simplistic and got perfect score of 100.

roji replied:
@smitpatel I don't know what Acrolinx is or how it scores, but "where async query" is missing an indefinite article, and using "avoid" twice sentence after sentence is problematic stylistically.

smitpatel replied:
This is outdated diff.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
 # Client vs. Server Evaluation
 
-Entity Framework Core supports parts of the query being evaluated on the client and parts of it being pushed to the database. It is up to the database provider to determine which parts of the query will be evaluated in the database.
+To evaluate query on server as much as possible, Entity Framework Core converts parts of the query to parameters, which it can evaluate on client side. Rest of the query with generated parameters is given to database providers to determine what should be equivalent database query to do evaluation on the server. EF Core supports partial client evaluation in top-level projection (essentially, the l ...
```
I don't see much value in specifying performance. This is doc about EF behavior. We can mention but not describe.

smitpatel replied:
Also there are queries where client eval would be more performant. :trollface:

smitpatel replied:
Change it little bit to simplify.

roji replied:
> I don't see much value in specifying performance

??

Performance is one of the two (extremely) major reasons we don't have client evaluation anymore... It seems critical for the understanding of what client vs. server evaluation is?

smitpatel replied:
I don't see value in specifying this
> This is important for performance: the server performs the work, possibly using indexes for efficiency, and only sends the results back to the client, minimizing network bandwidth needs.

Anyone using databases understands that. The intent of this page is not to empasize that in my mind.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+## Previous versions
 
-By default, EF Core will log a warning when client evaluation is performed. See [Logging](../miscellaneous/logging.md) for more information on viewing logging output. 
+The following section applies to EF Core versions before 3.0.
```
We need to determine the formatting on how to describe behavior in previous versions. I am keeping this as is for now.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+EF Core supported client evaluation in any part query that's why queries similar to one posted in [Unsupported client evaluation](#unsupported-client-evaluation) section worked correctly. But it caused performance issue without users noticing. To notify users that query was being evaluated on client, EF Core logged client evaluation warning. For more information on viewing logging output, see [Lo ...
 
-You can change the behavior when client evaluation occurs to either throw or do nothing. This is done when setting up the options for your context - typically in `DbContext.OnConfiguring`, or in `Startup.cs` if you are using ASP.NET Core.
+Optionally EF Core allowed to change default behavior to throw exception or do nothing when doing client evaluation except for projection. The exception throwing would make it similar to the behavior in 3.0. To change the behavior, you need to configure warnings while setting up the options for your context - typically in `DbContext.OnConfiguring`, or in `Startup.cs` if you're using ASP.NET Core.
```
There is also ignoring warnings rather than logging which above paragraph does not cover. ConfigureWarnings has Throw/Log/Ignore options.

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
-public static string StandardizeUrl(string url)
-{
-    url = url.ToLower();
+In the following example, a helper method is used to standardize URLs for blogs, which are returned from a SQL Server database. Since the SQL Server provider has no insight into how this method is implemented, it isn't possible to translate it into SQL. All other aspects of the query are evaluated in the database, but passing the returned `URL` through this method is done on the client.
```
Title makes it clear that it is in top-level projection.

smitpatel commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+---
+# Complex Query Operators
+
+Language Integrated Query (LINQ) contains many complex operators, which combine multiple data sources or does complex processing. Not all LINQ operators have suitable translation in the server side. Sometimes, query in one form translates to server but if written in different form doesn't translate even if the result is same. This page describes some of the complex operators and their supported v ...
```
It changes meaning of what we want to say.

smitpatel commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## Join
+
+LINQ Join operator allows you to connect two data sources based on the key selector on each source and generates a tuple of values when key matches. It naturally translates to `INNER JOIN` on relational databases. While LINQ Join has outer & inner key selectors, databases take a condition as join condition. So EF Core will generate join condition by comparing outer key selector to inner key selec ...
```
Not `The` before LINQ.

roji commented:
@smitpatel FYI am going to stop reviewing the responses to my PR, as 95% seem to be simple rejections with no explanation.

smitpatel commented:
@roji - I incorporated the suggestions just not as is because acrolinx complained about it. I have taken ideas from the suggestion to update the text to simplify it. The outdated items above and if you read the docs, I have updated the content. By no means I feel obligated to take all the suggestions as is.

cc: @ajcvickers @bricelam

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## GroupJoin
+
+LINQ GroupJoin operator allows you to connect two data sources in similar to Join but it creates a group of inner values for matching outer elements. Executing query like following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) don't support representing a collection of client-side objects, GroupJoin can't be translated to server in many cases. ...
```
```suggestion
LINQ's GroupJoin operator allows you to connect two data sources similar to Join, but it creates a group of inner values for matching outer elements. Executing a query like the following generates a result of `Blog` & `IEnumerable<Post>`. Since databases--especially relational databases--don't have a way to represent collections of client-side objects, GroupJoin can't be translated to the server in many cases. It requires you to get all of the data from the server to do GroupJoin without a special selector (first query below). But if the selector is limiting data being selected then fetching all of the data from the server may cause performance issues (second query below). That's why EF Core doesn't translate GroupJoin.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## SelectMany
+
+LINQ SelectMany operator allows you to enumerate over a collection selector for each outer element and generate tuples of value from each data source. In a way, it's a join but without any condition so every outer element is connected with element from collection source. Depending on how the collection selector is related to the outer data source, SelectMany can translate to various different que ...
```
```suggestion
LINQ's SelectMany operator allows you to enumerate over a collection selector for each outer element and generate tuples of values from each data source. In a way, it's a join but without any condition so every outer element is connected with an element from the collection source. Depending on how the collection selector is related to the outer data source, SelectMany can translate into various different queries on the server side.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+### Collection selector references outer in where clause
+
+When the collection selector has a where clause, which references the outer element, then EF Core translates it to a database join and uses the predicate as the join condition. Normally it arises when using collection navigation on the outer element as the collection selector. If the collection is empty for outer element, then no results would be generated for that outer element. But if `DefaultI ...
```
```suggestion
When the collection selector has a where clause, which references the outer element, then EF Core translates it to a database join and uses the predicate as the join condition. Normally this arises when using collection navigation on the outer element as the collection selector. If the collection is empty for an outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with a default value of the inner element. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+CROSS JOIN [Posts] AS [p]
+```
+
+### Collection selector references outer in where clause
````
```suggestion
### Collection selector references outer in a where clause
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
+```
+
+### Collection selector references outer in non-where case
````
```suggestion
### Collection selector references outer in a non-where case
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+### Collection selector references outer in non-where case
+
+When the collection selector has reference to the outer element, which isn't in where clause as the case above, it doesn't translate to database join. That's why we need to evaluate the collection selector for each outer element. It translates to `APPLY` operation in many relational databases. If the collection is empty for outer element, then no results would be generated for that outer element. ...
```
```suggestion
When the collection selector references the outer element, which isn't in a where clause (as the case above), it doesn't translate to a database join. That's why we need to evaluate the collection selector for each outer element. It translates to `APPLY` operations in many relational databases. If the collection is empty for an outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with a default value of the inner element. Because of this distinction, these kinds of queries translate to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. Certain databases like SQLite don't support `APPLY` operators so this kind of query may not be translated.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## GroupBy
+
+LINQ GroupBy operator creates result of type `IGrouping<TKey, TElement>` where `TKey` and `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over using any LINQ operator after grouping. Since no database structure can represent `IGrouping`, Group By operator has no translation in most cases. When an aggregate operator is appli ...
```
```suggestion
LINQ GroupBy operators create a result of type `IGrouping<TKey, TElement>` where `TKey` and `TElement` could be any arbitrary type. Furthermore, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over it using any LINQ operator after the grouping. Since no database structure can represent an `IGrouping`, Group By operator has no translation in most cases. When an aggregate operator is applied to each group, which returns a scalar, it can be translated to SQL `GROUP BY` in relational databases. SQL's `GROUP BY` is restrictive too. It requires you to group only by scalar values. The projection can only contain grouping key columns or any aggregate applied over a column. EF Core identifies this pattern and translates it to the server, as in the following example:
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+GROUP BY [p].[AuthorId]
+```
+
+EF Core also translates queries where an aggregate operator on the grouping appears in Where or OrderBy (or other ordering) LINQ operators. It uses `HAVING` clause in SQL for where clause. The query part before applying GroupBy operator can be any complex query as long as it can be translated to server. Further, once you apply aggregate operators on a grouping query to remove groupings from resul ...
````
```suggestion
EF Core also translates queries where an aggregate operator on the grouping appears in a Where or OrderBy (or other ordering) LINQ operator. It uses `HAVING` clauses in SQL for the where clause. The part of the query before applying the GroupBy operator can be any complex query as long as it can be translated to server. Furthermore, once you apply aggregate operators on a grouping query to remove groupings from the resulting source, you can compose on top of it like any other query.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+ORDER BY [p].[AuthorId]
+```
+
+The aggregate operators EF Core supports are following
````
```suggestion
The aggregate operators EF Core supports are as follows
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
```diff
@@ -0,0 +1,128 @@
+
+## Left Join
+
+While Left Join isn't a LINQ operator, relational database has concept of Left Join which is frequently used in queries. A particular pattern in LINQ query gives same result as `LEFT JOIN` on server. EF Core identifies such pattern and generate equivalent `LEFT JOIN` on server side. The pattern involves creating GroupJoin between both the data sources and then flattening out the grouping by using ...
```
```suggestion
While Left Join isn't a LINQ operator, relational databases have the concept of a Left Join which is frequently used in queries. A particular pattern in LINQ queries gives the same result as a `LEFT JOIN` on the server. EF Core identifies such patterns and generates the equivalent `LEFT JOIN` on the server side. The pattern involves creating a GroupJoin between both the data sources and then flattening out the grouping by using the SelectMany operator with DefaultIfEmpty on the grouping source to match null when the inner doesn't have a related element. The following example shows what that pattern looks like and what it generates.
```

bricelam commented on entity-framework/core/querying/complex-query-operators.md:
````diff
@@ -0,0 +1,128 @@
+LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
+```
+
+Above pattern creates complex structure in expression tree. Because of that EF Core requires you to flatten out grouping results out of GroupJoin operator in immediate step. Even if the GroupJoin-DefaultIfEmpty-SelectMany is used but in different pattern, we may not identify as Left Join.
````
```suggestion
The above pattern creates a complex structure in the expression tree. Because of that, EF Core requires you to flatten out the grouping results of the GroupJoin operator in a step immediately following the operator. Even if the GroupJoin-DefaultIfEmpty-SelectMany is used but in a different pattern, we may not identify it as a Left Join.
```

bricelam commented on entity-framework/core/querying/index.md:
```diff
@@ -1,10 +1,32 @@
+> [!TIP]
+> You can view this article's [sample](https://github.com/aspnet/EntityFramework.Docs/tree/master/samples/core/Querying) on GitHub.
+
+Following shows a few examples to achieve common tasks with Entity Framework Core.
```
```suggestion
The following shows a few examples of how to achieve common tasks with Entity Framework Core.
```

smitpatel commented on entity-framework/core/querying/async.md:
```diff
@@ -1,21 +1,21 @@
+> EF Core doesn't support multiple parallel operations being run on the same context instance. You should always wait for an operation to complete before beginning the next operation. This is typically done by using the `await` keyword on each async operation.
 
-Entity Framework Core provides a set of asynchronous extension methods that can be used as an alternative to the LINQ methods that cause a query to be executed and results returned. Examples include `ToListAsync()`, `ToArrayAsync()`, `SingleAsync()`, etc. There are not async versions of LINQ operators such as `Where(...)`, `OrderBy(...)`, etc. because these methods only build up the LINQ expressi ...
+Entity Framework Core provides a set of async extension methods similar to the LINQ methods, which execute a query and return results. Examples include `ToListAsync()`, `ToArrayAsync()`, `SingleAsync()`. There are no async versions of some LINQ operators such as `Where(...)`, `OrderBy(...)`, because these methods only build up the LINQ expression tree and don't cause the query to be executed in t ...
```
There are also other methods. How should we describe that?

smitpatel replied:
Converted to
> some LINQ operators such as `Where(...)` or `OrderBy(...)`

smitpatel replied:
Still 100 😉

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
 # Client vs. Server Evaluation
 
-Entity Framework Core supports parts of the query being evaluated on the client and parts of it being pushed to the database. It is up to the database provider to determine which parts of the query will be evaluated in the database.
+As a general rule, Entity Framework Core attempts to evaluate query on server as much as possible. EF Core converts parts of the query to parameters, which it can evaluate on client side. Rest of the query with generated parameters is given to database providers to determine what should be equivalent database query to do evaluation on the server. EF Core supports partial client evaluation in top- ...
```
also found another missing `the`

smitpatel commented on entity-framework/core/querying/client-eval.md:
```diff
@@ -1,69 +1,62 @@
+> [!NOTE]
+> Prior to version 3.0, Entity Framework Core supported client evaluation anywhere in the query. For more information, see the [previous versions section](#previous-versions).
+
+## Client evaluation in top-level projection
```
Went with the since there is only one top-level projection. It's not incorrect just bit ambiguous.
